contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,254 @@
+
+    /**
+     * @notice Initialize a new StableJoeStaking contract
+     * @dev This contract needs to receive an ERC20 `_rewardToken` in order to distribute them
+     * (with JoeMakerV4 in our case)
+     * @param _rewardToken The address of the ERC-20 reward token
+     * @param _joe The address of the JOE token
+     */
+    function initialize(IERC20Upgradeable _rewardToken, IERC20Upgradeable _joe) external initializer {
+        __Ownable_init();
+
+        rewardToken = _rewardToken;
+        joe = _joe;
+
+        lastTokenBalance = _rewardToken.balanceOf(address(this)); /// @dev Added to be upgrade safe
+        PRECISION = 1e12; /// @dev initialized here to be upgrade safe
+        lastDailyFeeTimestamp = (block.timestamp / 1 days) * 1 days;
+    }
+
+    /**
+     * @notice Deposit JOE to sJOE for reward token allocation
+     * @param _amount The amount of JOE to deposit
+     */
+    function deposit(uint256 _amount) external {
+        updatePool();
+
+        UserInfo storage user = userInfo[msg.sender];
+        if (user.amount > 0) {
+            uint256 pending = user.amount.mul(accTokenPerShare).div(PRECISION).sub(user.rewardDebt);
+            safeTokenTransfer(msg.sender, pending);
+        }
+        user.amount = user.amount.add(_amount);
+        user.rewardDebt = user.amount.mul(accTokenPerShare).div(PRECISION);
+
+        joe.safeTransferFrom(address(msg.sender), address(this), _amount);
+        emit Deposit(msg.sender, _amount);
+    }
+
+    /**
+     * @notice return the token per sec.
+     */
+    function tokenPerSec() external view returns (uint256) {
+        return tokenPerSecScaled.div(PRECISION);
+    }
+
+    /**
+     * @notice View function to see pending reward token on frontend
+     * @param _user The address of the user
+     * @return `_user`'s pending reward token
+     */
+    function pendingToken(address _user) external view returns (uint256) {
+        UserInfo storage user = userInfo[_user];
+        uint256 lpSupply = joe.balanceOf(address(this));
+        uint256 _accTokenPerShare = accTokenPerShare;
+        uint256 _multiplier;
+        uint256 _tokenReward;
+        uint256 _lastRewardTimestamp = lastRewardTimestamp;
+        uint256 _lastDailyFeeTimestamp = lastDailyFeeTimestamp;
+        uint256 _tokenPerSecScaled = tokenPerSecScaled;
+        uint256 _lastTokenBalance = lastTokenBalance;
+        if (block.timestamp > _lastDailyFeeTimestamp + 1 days) {
+            _multiplier = 1 days - (_lastRewardTimestamp % 1 days);
+
+            _tokenReward = _multiplier.mul(_tokenPerSecScaled);
+
+            _lastRewardTimestamp = _lastDailyFeeTimestamp + 1 days;
+            _lastDailyFeeTimestamp = (block.timestamp / 1 days) * 1 days;
+
+            uint256 _accruedFee = rewardToken.balanceOf(address(this)).sub(_lastTokenBalance);
+            _lastTokenBalance = _lastTokenBalance.add(_accruedFee);
+
+            _tokenPerSecScaled = _accruedFee.mul(PRECISION).div(1 days);
```

Comment:
it'd be great to add some comments in line here explaining the math to be explicit

Reply:
Indeed, I'll try to be more explicit